back-end/helpers/get/transcript_data_grouper.js

```diff
@@ -0,0 +1,159 @@
+
+    static #IntentDao;
+
+    constructor(transcript, project_id) {
+        this.transcript = transcript;
+        this.project_id = project_id;
+    }
+
+    static setIntentDao(dao){
+        if(dao instanceof IntentInterface){
+            this.#IntentDao = dao;
+        } else {
+            throw new Error("not an IntentInterface");
+        }
+    }
+
+
+    async group(intent) {
+        this.intent = intent;
+        for (const comparedIntent of this.transcript) {
+            if (this.#generateSimilarityScore(comparedIntent) >= 0.9) {
+                try {
+                    await this.#mergeIntent(comparedIntent);
+                    return true;
+                } catch (e) {
+                    console.log(e.message)
+                    return false;
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * @param Object Formatted intent to compare to
+     * @return number in [0, 1]
+     */
+    #generateSimilarityScore(comparedIntent) {
+        const splitIntent1 = this.intent.question.split(" ");
+        const splitIntent2 = comparedIntent.question.split(" ");
+        this.differentWordsIndices = [];
+
+        // If the length is significantly different we do not want to merge
+        if (Math.abs(splitIntent1.length - splitIntent2.length) > 3) {
+            return 0;
+        }
+
+        // Set a current score and let it decrement by 1 * weight each time a word is different
+        const DECREMENTOR = 1;
+        let score = splitIntent1.length;
+
+        // Set the weight of how much each difference matters
+        let weight = 1;
+        if (splitIntent1.length < 10) {
+            weight = 0.5;
+        }
+
+        // We use index to give us a greater control of manipulation over both arrays
+        let index1 = 0;
+        let index2 = 0;
+
+        while (index1 < splitIntent1.length && index2 < splitIntent2.length) {
+            if (splitIntent1[index1] !== splitIntent2[index2] && splitIntent2 !== "[merged]") {
+                score -= weight * DECREMENTOR;
+                this.differentWordsIndices.push(index1);
+
+                // Attempt to not be off-by-1 index for names by identifying titles
+                if (titles.includes(splitIntent1[index1].toUpperCase())) {
+                    index1 += 1;
+                }
+                if (titles.includes(splitIntent2[index2].toUpperCase())) {
+                    index2 += 1;
+                }
+            }
+            index1 += 1;
+            index2 += 1;
+        }
+        return score / splitIntent1.length;
+    }
+
+    /**
+     * @param Object Formatted intent to merge to
+     * Merges 2 intents
+     */
+    async #mergeIntent(comparedIntent) {
+        if (comparedIntent.previousIntents) {
+            // Make deep copy of original intent to not create recursion
+            let original = JSON.parse(JSON.stringify(comparedIntent.previousIntents[0]));
```

Comment:
Why do we need to make a deep copy of the original? At most I feel like we should make a deep copy of one part of the original. Ideally, we can find a more efficient way to do this (this method is quite costly)

Reply:
Yeah I'm assuming we can make a deep copy of only a couple parts since when I first did it there was a recursion issue

Reply:
Okay so I checked and it's because we don't want to modify the first intent in previous intents (which is the original intent) when we're recalculating so we need a new object that's a copy.